Files_match_new (#6)
ferrous-systems/teaching-material2: src/assignments/files-match-result.md

# Files, Match and Result

In this exercise you will complete a number of steps to learn about
Error Handling. The final result will be a url parser that reads lines
from a text file and can distinguish the content between URLs and
non-urls.

## In this exercise, you will learn how to

-   handle occurring `Result`-types with `match` for basic error
    handling.

-   when to use the `.unwrap()` method.

-   propagate an error with the `?` operator

-   return the `Option`-type.

-   do some elementary file processing (opening, reading to buffer,
    counting, reading line by line).

-   navigate the Rust `stdlib` documentation

-   add external dependencies to your project

## Task

✅ Clone the teaching material repository at
[github.com/ferrous-systems/teaching-material](https://github.com/ferrous-systems/teaching-material).

[todo!] add correct location once it's definite.
✅ Fix the runtime error by correcting the file path. 

✅ Manually unwrap the `Result` type that is returned from the

>     File::open() with a match statement, so that the .unwrap() can be

Reviewer: ```suggestion
    `File::open()` with a match statement, so that the `.unwrap()` can be
```

    deleted.

> ✅ Move this manual unwrap to it’s own function.

Reviewer: ```suggestion
✅ Move this manual unwrap to its own function.
```

✅ Read the content of the file to a buffer and count the lines in a
    for loop.

✅ Filter out empty lines and print the non-empty ones.

> ✅ Write a function that parses each line and returns Some(url) if the line is a URL, and `None` if it is not. Use the Url crate [Url

Reviewer: ```suggestion
✅ Write a function that parses each line and returns `Some(url)` if the line is a URL, and `None` if it is not. Use the Url crate's [Url 
```

    Type](https://docs.rs/url/2.1.1/url/)

## Knowledge

### Option and Result

Both `Option` and `Result` are similar in a way. Both have two
variants, and depending on what those variants are, the program may
continue in a different way.

> The `Option` Type can have the variant `Some(<some other type>)` or

Reviewer: Uses `Some(<some other type>)` for Option and `Ok(t)` when talking about Result. These should use the same syntax, as they are the same kind of thing.

Suggest using Rust syntax, so:

> The `Option` type can have the variant `Some(T)`, where `T` is a type parameter that means *some type should go here, we'll decide which one later*.

`None`. It is used, when you have to handle optional values. For example
if you want to be able to leave a field of a struct empty, you assign the
`Option` type to it. If the field has a value, it is `Some(<value>)`, if
it is empty, it is `None`.

The variants of the `Result` type are `Ok(t)` and `Err(e)`. It is used to
handle errors. If an operation was successful, `Ok(t)` is returned. In
`Ok(t)`, `t` can be the empty tuple or a return value. In `Err(e)`, `e`
contains an error message that can be printed.

Both types can be used with the `match` keyword. The received value is
matched on patterns, each leads to the execution of a different
expression.

### How to use `match`

`match` is a way of control flow based on pattern matching. A pattern on

> the one left evaluates to an expression on the right side.

Reviewer: ```suggestion
the left results in the expression on the right side.
```

```rust
match VALUE {
    PATTERN => EXPRESSION,
    PATTERN => EXPRESSION,
    PATTERN => EXPRESSION,
}
```

Other then with if/else, every case has to be handled explicitly, at
least with a last catch all arm that uses a place holder:

```rust
match VALUE {
    PATTERN => EXPRESSION,
    PATTERN => EXPRESSION,
    _ => EXPRESSION,
}
```

There are different ways to use match:

The return values of the expression can be bound to a variable:

```rust
let return_value = match VALUE {
    // match arms that return a value or panic
}
```

In case of a `Result<T, E>`, match statements can be used to get to
the inner value.

> ```rust

Reviewer: This isn't valid Rust. Maybe something like:

```rust
let some_result: Result<File, Error> = do_operation();
match some_result {
    Ok(file) => {
        // In this block we have a new variable `file`, of type `File`
    }
    Err(error) => {
        // In this block we have a new variable `err`, of type `Error`
    }
}
```
